### ISSUE TITLE
SIMBAD Exception when identifier not found

### ISSUE THREAD

**Issue by rickyegeland** [astropy/astroquery#473]
When identifiers are not found in `astroquery.simbad.Simbad.query_object` or `query_objectids`, a warning is printed and execution continues, e.g.:

> /.../astroquery/simbad/core.py:844: UserWarning: Warning: The script line number 2 raised an error (recorded in the `errors` attribute of the result table): Identifier not found in the database : HD 115404A
>   (error.line, error.msg))

This is probably a common & serious enough problem to warrant raising specific exception type, e.g. `IdentifierNotFoundError`, so that clients can decide what to do.

**Comment by keflavich** [comment]
Good idea.   PRs welcome; this shouldn't be too hard to implement.

**Comment by gully** [comment]
This issue is related to a broader feature I would like to see.  

Feature:  The result of a query (optionally) contains a column for the user-input identifier.

One could accomplish this with a boolean kwarg in query_object() or maybe Simbad(), like "return_input=True".  Identifiers that were not found in the query can simply be NaN's all the way across the row in the resulting Table.  This would make it much easier to grapple with IdentifierNotFoundErrors.

Forgive me if these features exist and I simply have missed them, but I cannot find how to achieve this goal.  I am new to contributing, so I'm not as knowledgable as others here.  If this feature is deemed valuable, I would be open to attempting a PR, but it could be ugly.

Thank you for reviewing this issue.

**Comment by keflavich** [comment]
@gully - I'm not entirely sure I understand.  You want the returned SIMBAD table to contain an additional column that was the input query?  This is possible, but I don't quite get the motivation.

To achieve what I described, you could do something like:
```
from astropy import table
result = Simbad.query_object('M82')
result.add_column(table.Column(name='InputName', data=['M82']))
```
Is that what you're looking for?  

If you do a `query_region` query, you can get multiple returns (unlike a `query_object`) - would you just want the queried coordinates in that case?

**Comment by gully** [comment]
Yes, this is what I want, but for query_objects.  In that case the user input is a list of identifiers:

```python
from astropy import table
result = Simbad.query_objects(['M82', 'M31', 'junk-identifier' ])
```
In this case the result is a Table with only 2 rows.  I am suggesting a third row is added with just NaN values.  Does that make sense?  The benefit here is that if a user merely has a list of source names, he/she can then instantly know which ones failed by boolean masking.  This problem is especially acute when handing in, say, 2MASS identifiers, but the Simbad identifier is something else.  Then I cannot easily identify which sources in the input list were not found.

**Comment by gully** [comment]
Here is a slightly more illustrative example:

```python
result2 = Simbad.query_objects(['M82', 'M31', 'junk-identifier', '2MASSJ10463795-7736035','M101'])
print result2["MAIN_ID"].data
```
['M  82' 'M  31' 'HD  93828' 'M 101']

In this case I don't know whether 'junk-identifier' or '2MASSJ10463795-7736035' corresponds to 'HD 93828'.  The proposed feature would have this output:

['M  82' 'M  31' 'NaN' 'HD  93828' 'M 101']

The need for retaining the input as a column arises when wildcards are used, for example.  Namely, any time the length of the input and result differ, it is ambiguous which results correspond to which input identifier.

**Comment by keflavich** [comment]
I understand now, thanks.  Good point.  This shouldn't be too hard to implement.  If you want to give it a shot, I'm happy to help out by reviewing any submitted code and providing feedback.

**Comment by gully** [comment]
Ok, so I made a work-around for this, posted below, with an example in the docstring.  My strategy was to parse the raw error message content with `Simbad.last_parsed_result.error_raw`.

I tested it on two different raw error messages: "Unrecognized identifier: " and "Identifier not found in the database : ".  They both worked fine.  I haven't thought about corner cases for the string parsing, though.  This solution is really just a work around / proof-of-concept.  A better solution would be to add a "keep_input=False" keyword argument in the `query_objects` function in Simbad core.py.

```python
def add_input_column_to_simbad_result(Sim, input_list, verbose=False):
    """Adds 'INPUT' column to the result of a Simbad query

    The motivation for this function is astroquery GitHub issue #473

    When you submit a batch query, you only get back the successes,
    often with different Simbad default names than the input list.
    This discrepancy hinders downstream Table join operations.

    Example:

        import astroquery.simbad
        input_list = ['M82', 'M31', 'junk-identifier1', 'junk-identifier2',
                      '2MASSJ10463795-7736035','M101']
        result = Simbad.query_objects(input_list)
        print result["MAIN_ID"].data

        ['M  82' 'M  31' 'HD  93828' 'M 101']

        result_fix = add_input_column_to_simbad_result(Simbad, input_list)
        print result_fix["INPUT"].data

        ['M82' 'M31' '2MASSJ10463795-7736035' 'M101']

    Args:
      Sim (SimbadClass): The Simbad class used to make the 
      input_list (list): The same list that was used to perform the query.
      verbose (boolean): Prints counts of successes/failures if True.

    Returns:
      Simbad Table with an additional column "INPUT"

    """
    error_string = Sim.last_parsed_result.error_raw
    fails = []

    for error in error_string.split("\n"):
        start_loc = error.rfind(":")+2
        fail = error[start_loc:]
        fails.append(fail)

    successes = [s for s in input_list if s not in fails]
    if verbose:
        out_message = "There were {} successful Simbad matches and {} failures."
        print out_message.format(len(successes), len(fails))
    
    Sim.last_parsed_result.table["INPUT"] = successes

    return Sim.last_parsed_result.table
```

**Comment by keflavich** [comment]
@gully - This looks pretty good, though there are a few things to clean up and we need to write a test for this.  Can you submit it as a pull request?  If you're not comfortable with git, you can do this through github by opening `simbad/core.py` and clicking the `edit` button (the pencil icon).  Then just add your code in the appropriate place, say just below `query_objects_async`.  Since it will be part of a class, the first argument should be `self` instead of `Sim`. 

The docstring also needs to conform to the numpy standard, which you can see in some of the other docstrings in `core.py`

**Comment by gully** [comment]
Roger that, thanks for reviewing the code snippet above.  I will make the requested changes and submit it as a Pull Request.  :smile:

**Comment by keflavich** [comment]
Closed by #1602
